Filebeat - panic while using multiline but give certain configure error

My OS version is CentOS Linux release 7.2.1511 (Core)
My Filebeat version is  filebeat version 5.2.1 (amd64), libbeat 5.2.1

I can produce a kernel panic for Filebeat when I specific a multiline setting like this

```
  ### Multiline options

  # Mutiline can be used for log messages spanning multiple lines. This is common
  # for Java Stack Traces or C-Line Continuation

  # The regexp Pattern that has to be matched. The example pattern matches all lines starting with [
#  multiline.pattern: ^\[
#  multiline.pattern: ^[^\[]|^\[.* SystemErr[\s\t]+R[\s\t]+at
  # Defines if the pattern set under pattern should be negated or not. Default is false.
#  multiline.negate: true
#  multiline.negate: false

  # Match can be set to "after" or "before". It is used to define if lines should be append to a pattern
  # that was (not) matched before or after or as long as a pattern is not matched based on negate.
  # Note: After is the equivalent to previous and before is the equivalent to to next in Logstash
  multiline.match: after

```

```
[kube@tivsun15 10.134.7.236]$ /usr/share/filebeat/bin/filebeat -path.home /usr/share/filebeat -path.config /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/filebeat -path.data /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/lib/filebeat -path.logs /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/log/filebeat etc_logstash_agent -e start
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5df0c6]

goroutine 45 [running]:
panic(0x8f44a0, 0xc420010090)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
regexp.(*Regexp).get(0x0, 0xc42034b528)
        /usr/local/go/src/regexp/regexp.go:208 +0x26
regexp.(*Regexp).doExecute(0x0, 0x0, 0x0, 0xc42032ea80, 0x5a, 0x5b, 0x0, 0x0, 0x0, 0x0, ...)
        /usr/local/go/src/regexp/exec.go:416 +0x40
regexp.(*Regexp).Match(0x0, 0xc42032ea80, 0x5a, 0x5b, 0x5a)
        /usr/local/go/src/regexp/regexp.go:421 +0x8c
github.com/elastic/beats/filebeat/harvester/reader.genPatternMatcher.func1(0xc4200532d0, 0x68, 0x69, 0xc42032ea80, 0x5a, 0x5b, 0x5b)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/multiline.go:307 +0xa3
github.com/elastic/beats/filebeat/harvester/reader.(*Multiline).readNext(0xc42000a300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/multiline.go:189 +0x8b7
github.com/elastic/beats/filebeat/harvester/reader.(*Multiline).readFirst(0xc42000a300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc42032d468, ...)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/multiline.go:132 +0x167
github.com/elastic/beats/filebeat/harvester/reader.(*Multiline).Next(0xc42000a300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420044cb0, ...)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/multiline.go:108 +0x62
github.com/elastic/beats/filebeat/harvester/reader.(*Limit).Next(0xc42036a7a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/limit.go:17 +0x66
github.com/elastic/beats/filebeat/harvester.(*Harvester).Harvest(0xc420334b60, 0xc07180, 0xc42036a7a0)
        /go/src/github.com/elastic/beats/filebeat/harvester/log.go:93 +0x24d
github.com/elastic/beats/filebeat/prospector.(*Prospector).startHarvester.func1(0xc420336000, 0xc420334b60, 0xc07180, 0xc42036a7a0)
        /go/src/github.com/elastic/beats/filebeat/prospector/prospector.go:248 +0x65
created by github.com/elastic/beats/filebeat/prospector.(*Prospector).startHarvester
        /go/src/github.com/elastic/beats/filebeat/prospector/prospector.go:249 +0x2e4
[kube@tivsun15 10.134.7.236]$
```

I agree the above setting is incorrect. But the expected behavior is exit with an error message, like below. A kernel panic due to an incorrect configuration must be a problem.

```
/usr/share/filebeat/bin/filebeat -path.home /usr/share/filebeat -path.config /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/filebeat -path.data /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/lib/filebeat -path.logs /home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/log/filebeat etc_logstash_agent -e start
Exiting: Error in initing prospector: missing required field accessing 'filebeat.prospectors.0.multiline.match' (source:'/home/kube/Softlayer_Logs/10.134.7.236/etc_logstash_agent/filebeat/filebeat.yml')
```

You can produce above error message by change filebeat.yml in this way. This time, the application exited as expected with an error message

```
  ### Multiline options

  # Mutiline can be used for log messages spanning multiple lines. This is common
  # for Java Stack Traces or C-Line Continuation

  # The regexp Pattern that has to be matched. The example pattern matches all lines starting with [
#  multiline.pattern: ^\[
#  multiline.pattern: ^[^\[]|^\[.* SystemErr[\s\t]+R[\s\t]+at
  # Defines if the pattern set under pattern should be negated or not. Default is false.
  multiline.negate: true
#  multiline.negate: false

  # Match can be set to "after" or "before". It is used to define if lines should be append to a pattern
  # that was (not) matched before or after or as long as a pattern is not matched based on negate.
  # Note: After is the equivalent to previous and before is the equivalent to to next in Logstash
#  multiline.match: after

```

--- a/filebeat/harvester/reader/multiline.go
+++ b/filebeat/harvester/reader/multiline.go
@@ -67,7 +67,7 @@ func NewMultiline(
 		return nil, fmt.Errorf("unknown matcher type: %s", config.Match)
 	}
 
-	matcher, err := matcherType(config.Pattern)
+	matcher, err := matcherType(*config.Pattern)
 	if err != nil {
 		return nil, err
 	}

--- a/filebeat/harvester/reader/multiline_config.go
+++ b/filebeat/harvester/reader/multiline_config.go
@@ -9,14 +9,15 @@ import (
 
 type MultilineConfig struct {
 	Negate       bool           `config:"negate"`
-	Match        string         `config:"match"       validate:"required"`
+	Match        string         `config:"match" validate:"required"`
 	MaxLines     *int           `config:"max_lines"`
-	Pattern      match.Matcher  `config:"pattern"`
-	Timeout      *time.Duration `config:"timeout"     validate:"positive"`
+	Pattern      *match.Matcher `config:"pattern" validate:"required"`
+	Timeout      *time.Duration `config:"timeout" validate:"positive"`
 	FlushPattern *match.Matcher `config:"flush_pattern"`
 }
 
 func (c *MultilineConfig) Validate() error {
+
 	if c.Match != "after" && c.Match != "before" {
 		return fmt.Errorf("unknown matcher type: %s", c.Match)
 	}

--- a/filebeat/harvester/reader/multiline_test.go
+++ b/filebeat/harvester/reader/multiline_test.go
@@ -24,9 +24,10 @@ func (p bufferSource) Stat() (os.FileInfo, error) { return nil, errors.New("unkn
 func (p bufferSource) Continuable() bool          { return false }
 
 func TestMultilineAfterOK(t *testing.T) {
+	pattern := match.MustCompile(`^[ \t] +`) // next line is indented by spaces
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern: match.MustCompile(`^[ \t] +`), // next line is indented by spaces
+			Pattern: &pattern,
 			Match:   "after",
 		},
 		2,
@@ -36,9 +37,11 @@ func TestMultilineAfterOK(t *testing.T) {
 }
 
 func TestMultilineBeforeOK(t *testing.T) {
+	pattern := match.MustCompile(`\\$`) // previous line ends with \
+
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern: match.MustCompile(`\\$`), // previous line ends with \
+			Pattern: &pattern,
 			Match:   "before",
 		},
 		2,
@@ -48,9 +51,11 @@ func TestMultilineBeforeOK(t *testing.T) {
 }
 
 func TestMultilineAfterNegateOK(t *testing.T) {
+	pattern := match.MustCompile(`^-`) // first line starts with '-' at beginning of line
+
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern: match.MustCompile(`^-`), // first line starts with '-' at beginning of line
+			Pattern: &pattern,
 			Negate:  true,
 			Match:   "after",
 		},
@@ -61,9 +66,11 @@ func TestMultilineAfterNegateOK(t *testing.T) {
 }
 
 func TestMultilineBeforeNegateOK(t *testing.T) {
+	pattern := match.MustCompile(`;$`) // last line ends with ';'
+
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern: match.MustCompile(`;$`), // last line ends with ';'
+			Pattern: &pattern,
 			Negate:  true,
 			Match:   "before",
 		},
@@ -75,10 +82,11 @@ func TestMultilineBeforeNegateOK(t *testing.T) {
 
 func TestMultilineAfterNegateOKFlushPattern(t *testing.T) {
 	flushMatcher := match.MustCompile(`EventEnd`)
+	pattern := match.MustCompile(`EventStart`)
 
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern:      match.MustCompile(`EventStart`),
+			Pattern:      &pattern,
 			Negate:       true,
 			Match:        "after",
 			FlushPattern: &flushMatcher,
@@ -92,10 +100,11 @@ func TestMultilineAfterNegateOKFlushPattern(t *testing.T) {
 
 func TestMultilineAfterNegateOKFlushPatternWhereTheFirstLinesDosentMatchTheStartPattern(t *testing.T) {
 	flushMatcher := match.MustCompile(`EventEnd`)
+	pattern := match.MustCompile(`EventStart`)
 
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern:      match.MustCompile(`EventStart`),
+			Pattern:      &pattern,
 			Negate:       true,
 			Match:        "after",
 			FlushPattern: &flushMatcher,
@@ -108,9 +117,10 @@ func TestMultilineAfterNegateOKFlushPatternWhereTheFirstLinesDosentMatchTheStart
 }
 
 func TestMultilineBeforeNegateOKWithEmptyLine(t *testing.T) {
+	pattern := match.MustCompile(`;$`) // last line ends with ';'
 	testMultilineOK(t,
 		MultilineConfig{
-			Pattern: match.MustCompile(`;$`), // last line ends with ';'
+			Pattern: &pattern,
 			Negate:  true,
 			Match:   "before",
 		},

--- a/filebeat/tests/system/test_multiline.py
+++ b/filebeat/tests/system/test_multiline.py
@@ -336,3 +336,19 @@ SetAdCodeMiddleware.default_ad_code route """
         output = self.read_output_json()
         output[0]["message"] = logentry1
         output[1]["message"] = logentry2
+
+    def test_invalid_config(self):
+        """
+        Test that filebeat errors if pattern is missing config
+        """
+        self.render_config_template(
+            path=os.path.abspath(self.working_dir + "/log/") + "*",
+            multiline=True,
+            match="after",
+        )
+
+        proc = self.start_beat()
+
+        self.wait_until(lambda: self.log_contains("missing required field accessing") == 1)
+
+        proc.check_kill_and_wait(exit_code=1)

